clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -7654,14 +7657,21 @@ void Sema::AddSYCLIntelFPGAMaxConcurrencyAttr(Decl *D,
       return;
     }
 
+    // Check to see if there's a duplicate attribute with different values
+    // already applied to the declaration.
     if (const auto *DeclAttr = D->getAttr<SYCLIntelFPGAMaxConcurrencyAttr>()) {
-      const auto *DeclExpr =
-          dyn_cast<ConstantExpr>(DeclAttr->getNThreadsExpr());
-      if (DeclExpr && ArgVal != DeclExpr->getResultAsAPSInt()) {
-        Diag(CI.getLoc(), diag::warn_duplicate_attribute) << CI;
-        Diag(DeclAttr->getLoc(), diag::note_previous_attribute);
+      // If the other attribute argument is instantiation dependent, we won't
+      // have converted it to a constant expression yet and thus we test
+      // whether this is a null pointer.
+      if (const auto *DeclExpr =
+              dyn_cast<ConstantExpr>(DeclAttr->getNThreadsExpr())) {
+        if (ArgVal != DeclExpr->getResultAsAPSInt()) {
+          Diag(CI.getLoc(), diag::warn_duplicate_attribute) << CI;
+          Diag(DeclAttr->getLoc(), diag::note_previous_attribute);
+        }
+        // Drop the duplicate attribute.
+        return;
```

Comment:
The 'return' statement has been moved so that we do not add duplicate attribute here:
Otherwise we will have duplicate attribute on AST:

```
|-OpenCLKernelAttr 0x55af79b232a0 <<invalid sloc>> Implicit
  |-ArtificialAttr 0x55af79b232f8 <<invalid sloc>> Implicit
  |-SYCLKernelAttr 0x55af79b234e8 <<invalid sloc>> Implicit
  |-AsmLabelAttr 0x55af79b23cb0 <<invalid sloc>> Implicit "_ZTSZZ3foovENKUlRN2cl4sycl7handlerEE_clES2_E13kernel_name_4"
  |-SYCLIntelFPGAMaxConcurrencyAttr 0x55af79b1ad60 <llvm/clang/test/SemaSYCL/max-concurrency-ast.cpp:90:16, col:40>
  | `-ConstantExpr 0x55af79b1ad40 <col:39> 'int'
  |   |-value: Int 3
  |   `-IntegerLiteral 0x55af79b1aab8 <col:39> 'int' 3
  `-SYCLIntelFPGAMaxConcurrencyAttr 0x55af79b1ade0 <line:91:16, col:40>
    `-ConstantExpr 0x55af79b1adc0 <col:39> 'int'
      |-value: Int 3
      `-IntegerLiteral 0x55af79b1aad8 <col:39> 'int' 3
```